edb/schema/pointers.py

```diff
@@ -1335,6 +1351,130 @@ def _alter_begin(
+            try:
+                p = cast(Pointer_T, schema.get_by_id(iid))
+                if not p.generic(schema) and p.get_owned(schema):
+                    lineage.append(p)
+            except errors.InvalidReferenceError:
+                pass
+
+        if is_owned:
+            # If the current pointer is explicitly declared, add it at
+            # the end of the lineage.
+            lineage.insert(0, scls)
+
+        status = self._validate_lineage(schema, lineage)
+
+        if status is LineageStatus.VALID:
+            return
+
+        if is_computable and is_owned:
+            # Overloading with a computable
+            raise errors.SchemaDefinitionError(
+                f'it is illegal for the computable '
+                f'{scls.get_verbosename(schema, with_parent=True)} '
+                f'to overload an existing '
+                f'{scls.get_schema_class_displayname()}',
+                context=self.source_context,
+            )
+        else:
+            if status is LineageStatus.MIXED:
+                raise errors.SchemaDefinitionError(
+                    f'it is illegal for the '
+                    f'{scls.get_verbosename(schema, with_parent=True)} '
+                    f'to extend both a computable and a non-computable '
+                    f'{scls.get_schema_class_displayname()}',
+                    context=self.source_context,
+                )
+            elif status is LineageStatus.MULTIPLE_COMPUTABLES:
+                raise errors.SchemaDefinitionError(
+                    f'it is illegal for the '
+                    f'{scls.get_verbosename(schema, with_parent=True)} '
+                    f'to extend more than one computable '
+                    f'{scls.get_schema_class_displayname()}',
+                    context=self.source_context,
+                )
+
+    def _validate_lineage(
+        self,
+        schema: s_schema.Schema,
+        lineage: List[Pointer_T],
+    ) -> LineageStatus:
+        if len(lineage) <= 1:
+            # Having at most 1 item in the lineage is always valid.
+            return LineageStatus.VALID
+
+        head, *rest = lineage
+
+        if not head.is_pure_computable(schema):
+            # The rest of the lineage must all be regular
+            if any(b.is_pure_computable(schema) for b in rest):
+                return LineageStatus.MIXED
+            else:
+                return LineageStatus.VALID
+
+        else:
+            # We have a computable with some non-empty lineage. Which
+            # could be valid only if this is some aliasing followed by
+            # regular pointers only.
+            prev_shortname = head.get_shortname(schema)
+            prev_is_comp = True
+            for b in rest:
+                cur_is_comp = b.is_pure_computable(schema)
+                cur_shortname = b.get_shortname(schema)
+                if prev_is_comp:
+                    # Computables cannot overload, but they can alias
```

Comment:
Could we check `from_alias` directly instead of guessing based on name equality?

Reply:
`from _alias` would mark the pointer that was derived, but the thing that we want to ensure is that the original pointer actually had a different name, so it was aliased and not overridden. I don't think we have anything labelling the original pointer in that way.